src/Microsoft.Net.Sdk.AnalyzerRedirecting/RedirectingAnalyzerAssemblyResolver.cs

```diff
@@ -0,0 +1,135 @@
+                    {
+                        continue;
+                    }
+
+                    string version = Path.GetFileName(versionDirectory);
+                    string majorVersionStr = version.IndexOf('.') is >= 0 and var index ? version.Substring(0, index) : version;
+                    if (!int.TryParse(majorVersionStr, out int majorVersion))
+                    {
+                        continue;
+                    }
+
+                    string analyzerName = Path.GetFileNameWithoutExtension(analyzerPath);
+                    string pathSuffix = analyzerPath.Substring(versionDirectory.Length + (EndsWithSlash(versionDirectory) ? 0 : 1));
+                    pathSuffix = Path.GetDirectoryName(pathSuffix);
+
+                    if (builder.TryGetValue(analyzerName, out var existing))
+                    {
+                        existing.Add((majorVersion, pathSuffix, analyzerPath));
+                    }
+                    else
+                    {
+                        builder.Add(analyzerName, new() { (majorVersion, pathSuffix, analyzerPath) });
+                    }
+                }
+            }
+        }
+
+        return builder.ToImmutable();
+    }
+
+    public string? RedirectPath(string fullPath)
+    {
+        if (AnalyzerMap.TryGetValue(Path.GetFileNameWithoutExtension(fullPath), out var analyzers))
+        {
+            foreach (var analyzer in analyzers)
+            {
+                var directoryPath = Path.GetDirectoryName(fullPath);
+                if (EndsWithIgnoringTrailingSlashes(directoryPath, analyzer.PathSuffix) &&
+                    hasMajorVersion(directoryPath, analyzer.PathSuffix, analyzer.MajorVersion))
+                {
+                    return analyzer.FullPath;
+                }
+            }
+        }
+
+        return null;
+
+        static bool hasMajorVersion(string directoryPath, string pathSuffix, int majorVersion)
+        {
+            // Find the version number in the directory path which is in the directory name before the path suffix.
+            int index = directoryPath.LastIndexOf(pathSuffix, StringComparison.OrdinalIgnoreCase);
+            if (index < 0)
+            {
+                return false;
+            }
+            string version = Path.GetFileName(Path.GetDirectoryName(directoryPath.Substring(0, index)));
+
+            // Check that the major version part matches.
+            return version.IndexOf('.') is >= 0 and var dotIndex &&
+                int.TryParse(version.Substring(0, dotIndex), out int versionMajor) &&
+                versionMajor == majorVersion;
+        }
+    }
+
+    public Assembly? ResolveAssembly(AssemblyName assemblyName, string assemblyOriginalDirectory)
+    {
+        if (AnalyzerMap.TryGetValue(assemblyName.Name, out var analyzers))
+        {
+            foreach (var analyzer in analyzers)
+            {
+                if (analyzer.MajorVersion == assemblyName.Version.Major &&
```

Comment:
Why are we trying to match on `AssemblyName.Version`?  I thought it was purely path based matching.  While our assemblies are often versioned consistently with the product - that's not something we need to couple to here.  Did you have a case that needed this?

Reply:
It seemed easier to use AssemblyName.Version than doing string parsing. That was actually before I added the RedirectPath API (which turned out to be needed so that roslyn can avoid the eager assembly load in some cases). After some more thought, I will remove this ResolveAssembly API and keep just the RedirectPath API.